File: src/openflight/iwr6843/live_selector.py
"""Deterministic reference for the firmware's bounded live range selector."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class SelectorParams:
    window_bins: int = 12
    max_jump_bins: int = 4
    max_misses: int = 2
    snr_q8: int = 768


@dataclass
class SelectorState:
    selected_bin: int = 0
    velocity_q8: int = 0
    misses: int = 0
    active: int = 0

    def copy(self) -> SelectorState:
        return SelectorState(
            self.selected_bin, self.velocity_q8, self.misses, self.active
        )


@dataclass(frozen=True)
class SelectorResult:
    candidate_bins: tuple[int, ...]
    candidate_power: tuple[int, ...]
    noise: int
    selected_bin: int
    window_start: int
    window_bins: int
    confidence_q8: int
    accepted: bool
    ambiguous: bool


def _window_start(selected_bin: int, window_bins: int, total_bins: int) -> int:
    return max(0, min(selected_bin - window_bins // 2, total_bins - window_bins))


def select_window(
    powers: list[int], params: SelectorParams, state: SelectorState
) -> SelectorResult:
    """Select at most two peaks and update a constant-velocity association."""
    if not powers or not 0 < params.window_bins <= len(powers):
        raise ValueError("invalid selector layout")
    noise = max(1, sum(powers) // len(powers))
    candidates = [
        index
        for index in range(1, len(powers) - 1)
        if powers[index] >= powers[index - 1]
        and powers[index] >= powers[index + 1]
        and powers[index] * 256 >= noise * params.snr_q8
    ]
    candidates.sort(key=lambda index: (-powers[index], index))
    candidates = candidates[:2]
    before_bin = state.selected_bin
    chosen: int | None = None
    if candidates and not state.active:
        chosen = candidates[0]
    elif candidates:
        predicted_q8 = state.selected_bin * 256 + state.velocity_q8
        eligible = [
            index
            for index in candidates
            if index >= state.selected_bin - 1
            and abs(index - state.selected_bin) <= params.max_jump_bins
            and abs(index * 256 - predicted_q8) <= params.max_jump_bins * 256
        ]
        if eligible:
            chosen = min(
                eligible,
                key=lambda index: (
                    abs(index * 256 - predicted_q8), -powers[index], index
                ),
            )
    if chosen is None:
        state.misses += 1
        selected = state.selected_bin
        state.velocity_q8 = 0
        if state.misses > params.max_misses:
            state.active = 0
            state.velocity_q8 = 0
    else:
        if state.active:
            state.velocity_q8 = int(
                (state.velocity_q8 + (chosen - before_bin) * 256) / 2
            )
        else:
            state.velocity_q8 = 0
        state.selected_bin = chosen
        state.misses = 0
        state.active = 1
        selected = chosen
    confidence = min(65535, powers[chosen] * 256 // noise) if chosen is not None else 0
    ambiguous = (
        len(candidates) == 2
        and powers[candidates[1]] * 256 >= powers[candidates[0]] * 230
    )
    return SelectorResult(
        candidate_bins=tuple(candidates),
        candidate_power=tuple(powers[index] for index in candidates),
        noise=noise,
        selected_bin=selected,
        window_start=_window_start(selected, params.window_bins, len(powers)),
        window_bins=params.window_bins,
        confidence_q8=confidence,
        accepted=chosen is not None,
        ambiguous=ambiguous,
    )

File: firmware/iwr6843/live_selector.c
#include "live_selector.h"

#include <stddef.h>
#include <string.h>

static uint32_t l3_abs_diff(int32_t left, int32_t right)
{
    return (uint32_t)(left >= right ? left - right : right - left);
}

int32_t l3_live_select(const uint32_t *powers, uint16_t nBins,
                       const L3LiveSelectorParams *params,
                       L3LiveSelectorState *state,
                       L3LiveSelectorResult *result)
{
    uint64_t sum = 0U;
    uint16_t bin;
    uint16_t candidates[2] = {0U, 0U};
    uint8_t count = 0U;
    int32_t chosen = -1;
    int32_t selected;
    int32_t predictedQ8;

    if (powers == NULL || params == NULL || state == NULL || result == NULL ||
        nBins < 3U || params->windowBins == 0U || params->windowBins > nBins) {
        return -1;
    }
    memset(result, 0, sizeof(*result));
    for (bin = 0U; bin < nBins; bin++) {
        sum += powers[bin];
    }
    result->noise = (uint32_t)(sum / nBins);
    if (result->noise == 0U) {
        result->noise = 1U;
    }
    for (bin = 1U; bin + 1U < nBins; bin++) {
        uint8_t position;
        if (powers[bin] < powers[bin - 1U] || powers[bin] < powers[bin + 1U] ||
            (uint64_t)powers[bin] * 256U <
                (uint64_t)result->noise * params->snrQ8) {
            continue;
        }
        position = 0U;
        while (position < count &&
               (powers[candidates[position]] > powers[bin] ||
                (powers[candidates[position]] == powers[bin] &&
                 candidates[position] < bin))) {
            position++;
        }
        if (position < 2U) {
            if (count < 2U) {
                count++;
            }
            if (count == 2U && position == 0U) {
                candidates[1] = candidates[0];
            }
            candidates[position] = bin;
        }
    }
    result->candidateCount = count;
    for (bin = 0U; bin < count; bin++) {
        result->candidateBins[bin] = candidates[bin];
        result->candidatePower[bin] = powers[candidates[bin]];
    }
    if (count > 0U && !state->active) {
        chosen = candidates[0];
    } else if (count > 0U) {
        uint32_t bestDistance = 0xFFFFFFFFU;
        predictedQ8 = (int32_t)state->selectedBin * 256 + state->velocityQ8;
        for (bin = 0U; bin < count; bin++) {
            int32_t candidateQ8 = (int32_t)candidates[bin] * 256;
            uint32_t distance = l3_abs_diff(candidateQ8, predictedQ8);
            if ((int32_t)candidates[bin] < (int32_t)state->selectedBin - 1 ||
                l3_abs_diff(candidates[bin], state->selectedBin) >
                    params->maxJumpBins ||
                distance > (uint32_t)params->maxJumpBins * 256U) {
                continue;
            }
            if (chosen < 0 || distance < bestDistance ||
                (distance == bestDistance &&
                 (powers[candidates[bin]] > powers[chosen] ||
                  (powers[candidates[bin]] == powers[chosen] &&
                   candidates[bin] < (uint16_t)chosen)))) {
                chosen = candidates[bin];
                bestDistance = distance;
            }
        }
    }
    if (chosen < 0) {
        state->misses++;
        selected = state->selectedBin;
        state->velocityQ8 = 0;
        if (state->misses > params->maxMisses) {
            state->active = 0U;
            state->velocityQ8 = 0;
        }
    } else {
        if (state->active) {
            state->velocityQ8 = (int16_t)((state->velocityQ8 +
                (chosen - state->selectedBin) * 256) / 2);
        } else {
            state->velocityQ8 = 0;
        }
        state->selectedBin = (int16_t)chosen;
        state->misses = 0U;
        state->active = 1U;
        selected = chosen;
        result->accepted = 1U;
        result->confidenceQ8 = (uint16_t)(
            ((uint64_t)powers[chosen] * 256U / result->noise) > 65535U
                ? 65535U
                : ((uint64_t)powers[chosen] * 256U / result->noise));
    }
    result->selectedBin = (uint16_t)selected;
    result->windowBins = params->windowBins;
    selected -= params->windowBins / 2U;
    if (selected < 0) {
        selected = 0;
    } else if (selected > (int32_t)(nBins - params->windowBins)) {
        selected = nBins - params->windowBins;
    }
    result->windowStart = (uint16_t)selected;
    result->ambiguous = count == 2U &&
        (uint64_t)powers[candidates[1]] * 256U >=
        (uint64_t)powers[candidates[0]] * 230U;
    return 0;
}
